### PR TITLE
Add support for visible_core_count for google_compute_instance resource

### PR DESCRIPTION
**Pull request by prateek2408**
Adding support for visible cores under google_compute_instance resource https://cloud.google.com/compute/docs/instances/customize-visible-cores

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: Added support for param `visible_core_count` in `google_compute_instance` and `google_compute_instance_template` under `advanced_machine_features`
```

### CODE REVIEW

**Review by slevenick — COMMENTED** [review]
This looks like it fails to build correctly. Are you able to build & run the tests locally?

**Comment by prateek2408** [comment]
@slevenick - How can I see why the build is failing in case of CI/CD. I am unable to see the project logs in GCP due to permission issues. 

Locally I tested and the only reason it can fail when validator is not built with **version=ga** parameter. Here is a snippet-

```
make test
# Skip integration tests in ./test/ using -short flag
GO111MODULE=on go test -short ./...
# github.com/GoogleCloudPlatform/terraform-validator/converters/google/resources
converters/google/resources/compute_instance_helpers.go:399:3: unknown field 'VisibleCoreCount' in struct literal of type compute.AdvancedMachineFeatures
converters/google/resources/compute_instance_helpers.go:410:59: AdvancedMachineFeatures.VisibleCoreCount undefined (type *compute.AdvancedMachineFeatures has no field or method VisibleCoreCount)
FAIL    github.com/GoogleCloudPlatform/terraform-validator/ancestrymanager [build failed]
FAIL    github.com/GoogleCloudPlatform/terraform-validator/cmd [build failed]
FAIL    github.com/GoogleCloudPlatform/terraform-validator/converters/google [build failed]
FAIL    github.com/GoogleCloudPlatform/terraform-validator/converters/google/resources [build failed]
FAIL    github.com/GoogleCloudPlatform/terraform-validator/test [build failed]
ok      github.com/GoogleCloudPlatform/terraform-validator/tfdata       (cached)
FAIL    github.com/GoogleCloudPlatform/terraform-validator/tfgcv [build failed]
ok      github.com/GoogleCloudPlatform/terraform-validator/tfplan       (cached)
make: *** [test] Error 2
```

**Comment by prateek2408** [comment]
@slevenick - Since the config is only present in beta I guarded the changes under the beta flag. I am able to build locally now without any flags.

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown [hunk 1]
```diff
@@ -577,7 +577,9 @@ The `specific_reservation` block supports:
 
 * `enable_nested_virtualization` (Optional) Defines whether the instance should have [nested virtualization](#on_host_maintenance) enabled. Defaults to false.
 
-* `threads_per_core` (Optional) he number of threads per physical core. To disable [simultaneous multithreading (SMT)](https://cloud.google.com/compute/docs/instances/disabling-smt) set this to 1.
+* `threads_per_core` (Optional) The number of threads per physical core. To disable [simultaneous multithreading (SMT)](https://cloud.google.com/compute/docs/instances/disabling-smt) set this to 1.
+
+* `visible_core_count` (Optional) The number of physical cores to expose to an instance. [visible cores info (VC)](https://cloud.google.com/compute/docs/instances/customize-visible-cores).
```
Comment (slevenick) [inline mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown:582]: This should say beta only

File: mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown [hunk 2]
```diff
@@ -414,6 +414,8 @@ specified, then this instance will have no external IPv6 Internet access. Struct
 
 * `threads_per_core` (Optional) he number of threads per physical core. To disable [simultaneous multithreading (SMT)](https://cloud.google.com/compute/docs/instances/disabling-smt) set this to 1.
 
+* `visible_core_count` (Optional) The number of physical cores to expose to an instance. [visible cores info (VC)](https://cloud.google.com/compute/docs/instances/customize-visible-cores).
```
Comment (slevenick) [inline mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown:417]: Beta only?

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb [hunk 3]
```diff
@@ -4742,6 +4742,7 @@ resource "google_compute_instance" "foobar" {
   advanced_machine_features {
 	threads_per_core = 1
 	enable_nested_virtualization = true
+	visible_core_count = 1
```
Comment (slevenick) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:4745]: Beta only?

File: mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb [hunk 4]
```diff
@@ -677,6 +677,13 @@ func resourceComputeInstanceTemplate() *schema.Resource {
 							ForceNew:    true,
 							Description: `The number of threads per physical core. To disable simultaneous multithreading (SMT) set this to 1. If unset, the maximum number of threads supported per core by the underlying processor is assumed.`,
 						},
+						"visible_core_count": {
```
Comment (slevenick) [inline mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb:680]: This should be beta only as well

**Review by slevenick — CHANGES_REQUESTED** [review]
You'll need to make sure this only appears in the beta version of the resource, as well as only in beta tests, and we need notes on the documentation that it is only in beta

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb (hunk 3 above)
Comment (prateek2408) [reply mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:4746]: Fixed

File: mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb (hunk 4 above)
Comment (prateek2408) [reply mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb:680]: Fixed

File: mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown (hunk 2 above)
Comment (prateek2408) [reply mmv1/third_party/terraform/website/docs/r/compute_instance.html.markdown]: Fixed

File: mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown (hunk 1 above)
Comment (prateek2408) [reply mmv1/third_party/terraform/website/docs/r/compute_instance_template.html.markdown]: Fixed

**Comment by slevenick** [comment]
This is still being added to the GA provider, when it should not be: https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6414-old..auto-pr-6414

Can you add the version guards around all except for the documentation?

**Comment by prateek2408** [comment]
> This is still being added to the GA provider, when it should not be: https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6414-old..auto-pr-6414
> 
> Can you add the version guards around all except for the documentation?

Fixed it, seems some files were missed earlier

File: mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb [hunk 5]
```diff
@@ -677,6 +677,15 @@ func resourceComputeInstanceTemplate() *schema.Resource {
 							ForceNew:    true,
 							Description: `The number of threads per physical core. To disable simultaneous multithreading (SMT) set this to 1. If unset, the maximum number of threads supported per core by the underlying processor is assumed.`,
 						},
+<% unless version == 'ga' -%>
+						"visible_core_count": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Computed:    false,
```
Comment (slevenick) [inline mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb:684]: You can remove Computed: false as it is the inferred value

File: mmv1/third_party/terraform/resources/resource_compute_instance.go.erb [hunk 6]
```diff
@@ -751,6 +751,14 @@ func resourceComputeInstance() *schema.Resource {
 							AtLeastOneOf: []string{"advanced_machine_features.0.enable_nested_virtualization","advanced_machine_features.0.threads_per_core"},
 							Description: `The number of threads per physical core. To disable simultaneous multithreading (SMT) set this to 1. If unset, the maximum number of threads supported per core by the underlying processor is assumed.`,
 						},
+<% unless version == 'ga' -%>
+						"visible_core_count": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							AtLeastOneOf: []string{"advanced_machine_features.0.enable_nested_virtualization","advanced_machine_features.0.visible_core_count"},
```
Comment (slevenick) [inline mmv1/third_party/terraform/resources/resource_compute_instance.go.erb:758]: Can you add the visible_core_count to the AtLeastOneOf of the above fields as well? Just for consistency's sake

File: mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb [hunk 7]
```diff
@@ -4742,6 +4742,9 @@ resource "google_compute_instance" "foobar" {
   advanced_machine_features {
 	threads_per_core = 1
 	enable_nested_virtualization = true
+	<% unless version == 'ga' -%>
+	visible_core_count = 1
```
Comment (slevenick) [inline mmv1/third_party/terraform/tests/resource_compute_instance_test.go.erb:4746]: Spacing here is a bit goofy in the generated test. Looks like extra tabs on the version guard lines are causing this

**Review by slevenick — CHANGES_REQUESTED** [review]
Couple minor changes needed, but overall looks good

File: mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb (hunk 5 above)
Comment (prateek2408) [reply mmv1/third_party/terraform/resources/resource_compute_instance_template.go.erb:684]: Done

File: mmv1/third_party/terraform/resources/resource_compute_instance.go.erb (hunk 6 above)
Comment (prateek2408) [reply mmv1/third_party/terraform/resources/resource_compute_instance.go.erb:758]: Fixed.

**Comment by prateek2408** [comment]
@slevenick - I incorporated the comments. I skipped adding visible_core_count in AtLeastOneOf as it is beta field. Adding this might cause the ga provided to get this  as well. 

Thoughts ?

**Review by slevenick — APPROVED** [review]
Looks good, thanks!

**Pull request merged**